man/fmtcheck.3
.\" Copyright (c) 2000 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This file was contributed to The NetBSD Foundation by Allen Briggs.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.\" $FreeBSD: /repoman/r/ncvs/src/lib/libc/gen/fmtcheck.3,v 1.9 2004/07/02 23:52:10 ru Exp $
.Dd October 16, 2002
.Os
.Dt FMTCHECK 3
.Sh NAME
.Nm fmtcheck
.Nd sanitizes user-supplied
.Xr printf 3 Ns -style
format string
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In stdio.h
.Ft const char *
.Fn fmtcheck "const char *fmt_suspect" "const char *fmt_default"
.Sh DESCRIPTION
The
.Fn fmtcheck
scans
.Fa fmt_suspect
and
.Fa fmt_default
to determine if
.Fa fmt_suspect
will consume the same argument types as
.Fa fmt_default
and to ensure that
.Fa fmt_suspect
is a valid format string.
.Pp
The
.Xr printf 3
family of functions cannot verify the types of arguments that they are
passed at run-time.
In some cases, like
.Xr catgets 3 ,
it is useful or necessary to use a user-supplied format string with no
guarantee that the format string matches the specified arguments.
.Pp
The
.Fn fmtcheck
was designed to be used in these cases, as in:
.Bd -literal -offset indent
printf(fmtcheck(user_format, standard_format), arg1, arg2);
.Ed
.Pp
In the check, field widths, fillers, precisions, etc.\& are ignored (unless
the field width or precision is an asterisk
.Ql *
instead of a digit string).
Also, any text other than the format specifiers
is completely ignored.
.Sh RETURN VALUES
If
.Fa fmt_suspect
is a valid format and consumes the same argument types as
.Fa fmt_default ,
then the
.Fn fmtcheck
will return
.Fa fmt_suspect .
Otherwise, it will return
.Fa fmt_default .
.Sh SECURITY CONSIDERATIONS
Note that the formats may be quite different as long as they accept the
same arguments.
For example,
.Qq Li "%p %o %30s %#llx %-10.*e %n"
is compatible with
.Qq Li "This number %lu %d%% and string %s has %qd numbers and %.*g floats (%n)" .
However,
.Qq Li %o
is not equivalent to
.Qq Li %lx
because
the first requires an integer and the second requires a long.
.Sh SEE ALSO
.Xr printf 3
.Sh BUGS
The
.Fn fmtcheck
function does not understand all of the conversions that
.Xr printf 3
does.


ChangeLog
2008-03-16  Guillem Jover  <guillem@debian.org>

	* include/vis.h: Remove UC Berkeley advertising clause, as per
	<ftp://ftp.cs.berkeley.edu/pub/4bsd/README.Impt.License.Change>.
	* src/vis.c: Likewise.
	* src/unvis.c: Likewise.
	* src/heapsort.c: Likewise.
	* src/linkaddr.c: Likewise.
	* man/fgetln.3: Likewise.

2007-12-18  Petr Salinger  <Petr.Salinger@seznam.cz>

	* include/bsd/queue.h: sync with current glibc variant on GNU/kFreeBSD

2006-11-08  Aurelien Jarno  <aurelien@aurel32.net>

	* src/link_addr.c: New file.
	* include/bsd/if_dl.h: Add link_addr.c declarations.
	* Makefile (LIB_SRCS): Add link_addr.c.
	* Versions: add link_addr and link_ntoa

2006-08-25  Petr Salinger  <Petr.Salinger@seznam.cz>

	* src/humanize_number.c: New file.
	* include/libutil.h: New file with humanize_number declarations.
	* Makefile (LIB_SRCS): Add humanize_number.c.
	* Versions: Add humanize_number.

2006-03-29  Robert Millan  <rmh@aybabtu.com>

	* src/heapsort.c: New file.
	* include/bsd/stdlib.h: Add heapsort declaration.
	* Makefile (LIB_SRCS): Add heapsort.c.
	* Versions: Add heapsort.

2006-03-19  Robert Millan  <rmh@aybabtu.com>

	* src/vis.c: New file.
	* src/unvis.c: Ditto.
	* include/vis.h: Add vis/unvis declarations.
	* Makefile (LIB_SRCS): Add vis.c/unvis.c.
	* Versions: Add vis/unvis functions.

2006-03-17  Robert Millan  <rmh@aybabtu.com>

	* src/progname.c: New file.
	* include/bsd/stdlib.h: Add {get,set}progname declarations.
	* Makefile (LIB_SRCS): Add 'progname.c'.
	* Versions: Add getprogname and setprogname.

2006-02-15  Guillem Jover  <guillem@debian.org>

	* include/bsd/cdefs.h: Disable __unused, it conflicts with a struct
	member in a Linux header.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* include/bsd/stdlib.h: Remove <sys/types.h> and <stdio.h>.
	* include/bsd/getopt.h: Protect header from multiple inclusion.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* include/bsd/bsd.h: Remove commented out macros.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* include/bsd/bsd.h (setproctitle, __dead2, __unused, __printflike)
	(__FBSDID): Moved to ...
	* include/bsd/cdefs.h: ... here.
	(__unused, __printflike) [__GNUC__]: Use proper __attribute__.
	* Makefile (LIB_INCLUDES): Add 'cdefs.h'.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* arc4random.c: Move to ...
	* src/arc4random.c: ... here.
	* bsd_getopt.c: Move to ...
	* src/bsd_getopt.c:: ... here.
	* err.c: Move to ...
	* src/err.c:: ... here.
	* fgetln.c: Move to ...
	* src/fgetln.c:: ... here.
	* fmtcheck.c: Move to ...
	* src/fmtcheck.c:: ... here.
	* inet_net_pton.c: Move to ...
	* src/inet_net_pton.c:: ... here.
	* strlcat.c: Move to ...
	* src/strlcat.c:: ... here.
	* strlcpy.c: Move to ...
	* src/strlcpy.c:: ... here.
	* md5c.c: Move to ...
	* src/md5c.c:: ... here.
	* md5.copyright: Move to ...
	* src/md5.copyright:: ... here.
	* Makefile (LIB_SRCS): Automatically prefix the src/ dir.

2006-02-14  Guillem Jover  <guillem@debian.org>

	* Makefile (LIB_SRCS): Line wrap.
	(LIB_INCLUDES): List only the files, automatically prefix the dir.
	(LIB_MANS): Likewise.

2006-02-13  Robert Millan  <rmh@aybabtu.com>

	* bsd_getopt.c: New file.
	* include/bsd/getopt.h: Likewise.
	* Makefile (LIB_SRCS): Add 'bsd_getopt.c'.
	(LIB_INCLUDES): Add 'include/bsd/getopt.h'.
	* Versions: Added bsd_getopt and optreset.

2006-02-10  Robert Millan  <rmh@aybabtu.com>

	* err.c: New file.
	* include/bsd/err.h: Likewise.
	* Makefile (LIB_SRCS): Add 'err.c'.
	(LIB_INCLUDES): Add 'include/bsd/err.h'.
	* Versions: Added errc, warnc, verrc and vwarn.

2005-12-19  Aurelien Jarno <aurel32@debian.org>

	* man/arc4random.3: New file.
	* man/fgetln.3: Likewise.
	* man/fmtcheck.3: Likewise.
	* man/strlcpy.3: Likewise.
	* fmtcheck.c:: Likewise.
	* include/stdlib.h: Added fmtcheck declaration.
	* Versions: Added fmtcheck.

2005-11-01  Guillem Jover  <guillem@debian.org>

	* ip_icmp.h: Move to ...
	* include/bsd/ip_icmp.h: ... here.
	* md5.h: Move to ...
	* include/bsd/md5.h: ... here.
	* bsd.h: Move to ...
	* include/bsd/bsd.h: ... here. Split contents ...
	* include/bsd/random.h: ... here ...
	* include/bsd/queue.h: ... here ...
	* include/bsd/string.h: ... and here.
	* md5c.c: Change md5.h include line.
	* Makefile (MK_CFLAGS): Include from the proper dir.

2005-08-03  Hector Garcia  <hector@debian.org>

	* Makefile: Fixed to place soft links instead in soname, and
	shared_so instead of copying files.
	* md5.h (MD5End, MD5File, MD5FileChunk, MD5Data): Commented out
	functions not available on 'md5c.c'.

2005-07-26  Guillem Jover  <guillem@debian.org>

	* Makefile (LIB_INCLUDES): Add 'md5.h'.

2005-07-26  Guillem Jover  <guillem@debian.org>

	* Versions (LIBBSD_0.0): Add MD5Init, MD5Update, MD5Pad and
	MD5Final.

2005-07-26  Hector Garcia  <hector@debian.org>

	* Makefile (LIB_SRCS): Add 'md5c.c'.
	* md5c.c: Changed include to 'md5.h'.
	* md5.h: Replaced with the system header 'sys/md5.h'.

2005-07-25  Guillem Jover  <guillem@debian.org>

	* fgetln.c (fgetln): Reindent, remove commented code and translate
	variable names to english.

2005-07-25  Guillem Jover  <guillem@debian.org>

	* Versions: New file.
	* Makefile: Add versioned symbols support.



src/fmtcheck.c
/*-
 * Copyright (c) 2000 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code was contributed to The NetBSD Foundation by Allen Briggs.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

/*	$NetBSD: fmtcheck.c,v 1.2 2000/11/01 01:17:20 briggs Exp $	*/
#include <sys/cdefs.h>
__FBSDID("$FreeBSD: /repoman/r/ncvs/src/lib/libc/gen/fmtcheck.c,v 1.8 2005/03/21 08:00:55 das Exp $");

#include <stdio.h>
#include <string.h>
#include <ctype.h>

/* __weak_reference(__fmtcheck, fmtcheck); */

enum __e_fmtcheck_types {
	FMTCHECK_START,
	FMTCHECK_SHORT,
	FMTCHECK_INT,
	FMTCHECK_LONG,
	FMTCHECK_QUAD,
	FMTCHECK_PTRDIFFT,
	FMTCHECK_SIZET,
	FMTCHECK_SHORTPOINTER,
	FMTCHECK_INTPOINTER,
	FMTCHECK_LONGPOINTER,
	FMTCHECK_QUADPOINTER,
	FMTCHECK_PTRDIFFTPOINTER,
	FMTCHECK_SIZETPOINTER,
#ifndef NO_FLOATING_POINT
	FMTCHECK_DOUBLE,
	FMTCHECK_LONGDOUBLE,
#endif
	FMTCHECK_STRING,
	FMTCHECK_WIDTH,
	FMTCHECK_PRECISION,
	FMTCHECK_DONE,
	FMTCHECK_UNKNOWN
};
typedef enum __e_fmtcheck_types EFT;

#define RETURN(pf,f,r) do { \
			*(pf) = (f); \
			return r; \
		       } /*NOTREACHED*/ /*CONSTCOND*/ while (0)

static EFT
get_next_format_from_precision(const char **pf)
{
	int		sh, lg, quad, longdouble, ptrdifft, sizet;
	const char	*f;

	sh = lg = quad = longdouble = ptrdifft = sizet = 0;

	f = *pf;
	switch (*f) {
	case 'h':
		f++;
		sh = 1;
		break;
	case 'l':
		f++;
		if (!*f) RETURN(pf,f,FMTCHECK_UNKNOWN);
		if (*f == 'l') {
			f++;
			quad = 1;
		} else {
			lg = 1;
		}
		break;
	case 'q':
		f++;
		quad = 1;
		break;
	case 't':
		f++;
		ptrdifft = 1;
		break;
	case 'z':
		f++;
		sizet = 1;
		break;
	case 'L':
		f++;
		longdouble = 1;
		break;
	default:
		break;
	}
	if (!*f) RETURN(pf,f,FMTCHECK_UNKNOWN);
	if (strchr("diouxX", *f)) {
		if (longdouble)
			RETURN(pf,f,FMTCHECK_UNKNOWN);
		if (lg)
			RETURN(pf,f,FMTCHECK_LONG);
		if (quad)
			RETURN(pf,f,FMTCHECK_QUAD);
		if (ptrdifft)
			RETURN(pf,f,FMTCHECK_PTRDIFFT);
		if (sizet)
			RETURN(pf,f,FMTCHECK_SIZET);
		RETURN(pf,f,FMTCHECK_INT);
	}
	if (*f == 'n') {
		if (longdouble)
			RETURN(pf,f,FMTCHECK_UNKNOWN);
		if (sh)
			RETURN(pf,f,FMTCHECK_SHORTPOINTER);
		if (lg)
			RETURN(pf,f,FMTCHECK_LONGPOINTER);
		if (quad)
			RETURN(pf,f,FMTCHECK_QUADPOINTER);
		if (ptrdifft)
			RETURN(pf,f,FMTCHECK_PTRDIFFTPOINTER);
		if (sizet)
			RETURN(pf,f,FMTCHECK_SIZETPOINTER);
		RETURN(pf,f,FMTCHECK_INTPOINTER);
	}
	if (strchr("DOU", *f)) {
		if (sh + lg + quad + longdouble + ptrdifft + sizet)
			RETURN(pf,f,FMTCHECK_UNKNOWN);
		RETURN(pf,f,FMTCHECK_LONG);
	}
#ifndef NO_FLOATING_POINT
	if (strchr("aAeEfFgG", *f)) {
		if (longdouble)
			RETURN(pf,f,FMTCHECK_LONGDOUBLE);
		if (sh + lg + quad + ptrdifft + sizet)
			RETURN(pf,f,FMTCHECK_UNKNOWN);
		RETURN(pf,f,FMTCHECK_DOUBLE);
	}
#endif
	if (*f == 'c') {
		if (sh + lg + quad + longdouble + ptrdifft + sizet)
			RETURN(pf,f,FMTCHECK_UNKNOWN);
		RETURN(pf,f,FMTCHECK_INT);
	}
	if (*f == 's') {
		if (sh + lg + quad + longdouble + ptrdifft + sizet)
			RETURN(pf,f,FMTCHECK_UNKNOWN);
		RETURN(pf,f,FMTCHECK_STRING);
	}
	if (*f == 'p') {
		if (sh + lg + quad + longdouble + ptrdifft + sizet)
			RETURN(pf,f,FMTCHECK_UNKNOWN);
		RETURN(pf,f,FMTCHECK_LONG);
	}
	RETURN(pf,f,FMTCHECK_UNKNOWN);
	/*NOTREACHED*/
}

static EFT
get_next_format_from_width(const char **pf)
{
	const char	*f;

	f = *pf;
	if (*f == '.') {
		f++;
		if (*f == '*') {
			RETURN(pf,f,FMTCHECK_PRECISION);
		}
		/* eat any precision (empty is allowed) */
		while (isdigit(*f)) f++;
		if (!*f) RETURN(pf,f,FMTCHECK_UNKNOWN);
	}
	RETURN(pf,f,get_next_format_from_precision(pf));
	/*NOTREACHED*/
}

static EFT
get_next_format(const char **pf, EFT eft)
{
	int		infmt;
	const char	*f;

	if (eft == FMTCHECK_WIDTH) {
		(*pf)++;
		return get_next_format_from_width(pf);
	} else if (eft == FMTCHECK_PRECISION) {
		(*pf)++;
		return get_next_format_from_precision(pf);
	}

	f = *pf;
	infmt = 0;
	while (!infmt) {
		f = strchr(f, '%');
		if (f == NULL)
			RETURN(pf,f,FMTCHECK_DONE);
		f++;
		if (!*f)
			RETURN(pf,f,FMTCHECK_UNKNOWN);
		if (*f != '%')
			infmt = 1;
		else
			f++;
	}

	/* Eat any of the flags */
	while (*f && (strchr("#'0- +", *f)))
		f++;

	if (*f == '*') {
		RETURN(pf,f,FMTCHECK_WIDTH);
	}
	/* eat any width */
	while (isdigit(*f)) f++;
	if (!*f) {
		RETURN(pf,f,FMTCHECK_UNKNOWN);
	}

	RETURN(pf,f,get_next_format_from_width(pf));
	/*NOTREACHED*/
}

__const char *
fmtcheck(const char *f1, const char *f2)
{
	const char	*f1p, *f2p;
	EFT		f1t, f2t;

	if (!f1) return f2;
	
	f1p = f1;
	f1t = FMTCHECK_START;
	f2p = f2;
	f2t = FMTCHECK_START;
	while ((f1t = get_next_format(&f1p, f1t)) != FMTCHECK_DONE) {
		if (f1t == FMTCHECK_UNKNOWN)
			return f2;
		f2t = get_next_format(&f2p, f2t);
		if (f1t != f2t)
			return f2;
	}
	return f1;
}
